go/cmd/server/logging/logging.go

```diff
@@ -0,0 +1,306 @@
+	columnNameToStatus := make(map[string][]int32)
+	columnNameToTimestamp := make(map[string][]int64)
+	entityNameToEntityValues := make(map[string][]*types.Value)
+
+	for _, l := range memoryBuffer.logs {
+		// EntityTypes maps an entity name to the specific type and also which index in the entityValues array it is
+		// e.g if an Entity Key is {driver_id, customer_id}, then the driver_id entitytype would be dtype=int64, index=0.
+		// It's in the order of the entities as given by the schema.
+		for idx, entityName := range fcoSchema.Entities {
+			if _, ok := entityNameToEntityValues[entityName]; !ok {
+				entityNameToEntityValues[entityName] = make([]*types.Value, 0)
+			}
+			entityNameToEntityValues[entityName] = append(entityNameToEntityValues[entityName], l.EntityValue[idx])
+		}
+
+		// Contains both fv and odfv feature value types => they are processed in order of how the appear in the featureService
+		for idx, featureName := range fcoSchema.Features {
+			// populate the proto value arrays with values from memory buffer in separate columns one for each feature name
+			if _, ok := columnNameToProtoValueArray[featureName]; !ok {
+				columnNameToProtoValueArray[featureName] = make([]*types.Value, 0)
+				columnNameToStatus[featureName] = make([]int32, 0)
+				columnNameToTimestamp[featureName] = make([]int64, 0)
+			}
+			columnNameToProtoValueArray[featureName] = append(columnNameToProtoValueArray[featureName], l.FeatureValues[idx])
+			columnNameToStatus[featureName] = append(columnNameToStatus[featureName], int32(l.FeatureStatuses[idx]))
+			columnNameToTimestamp[featureName] = append(columnNameToTimestamp[featureName], l.EventTimestamps[idx].AsTime().UnixNano()/int64(time.Millisecond))
+		}
+	}
+
+	fields := make([]arrow.Field, 0)
+	columns := make([]array.Interface, 0)
+	for _, entityName := range fcoSchema.Entities {
+		protoArr := entityNameToEntityValues[entityName]
+		if len(protoArr) == 0 {
+			break
+		}
+		valArrowArray, err := gotypes.ProtoValuesToArrowArray(protoArr, arrowMemory, len(columnNameToProtoValueArray))
+		if err != nil {
+			return nil, err
+		}
+		arrowType, err := gotypes.ValueTypeEnumToArrowType(fcoSchema.EntityTypes[entityName])
+		if err != nil {
+			return nil, err
+		}
+		fields = append(fields, arrow.Field{
+			Name: entityName,
+			Type: arrowType,
+		})
+		columns = append(columns, valArrowArray)
+	}
+
+	for _, featureName := range fcoSchema.Features {
+
+		protoArr := columnNameToProtoValueArray[featureName]
+		if len(protoArr) == 0 {
+			break
+		}
+		arrowArray, err := gotypes.ProtoValuesToArrowArray(protoArr, arrowMemory, len(columnNameToProtoValueArray))
+		if err != nil {
+			return nil, err
+		}
+
+		arrowType, err := gotypes.ValueTypeEnumToArrowType(fcoSchema.FeaturesTypes[featureName])
+
+		if err != nil {
+			return nil, err
+		}
+		fields = append(fields, arrow.Field{
+			Name: featureName,
+			Type: arrowType,
+		})
+		columns = append(columns, arrowArray)
+	}
+	schema := arrow.NewSchema(
```

Comment:
we should build here schema from feature service specs. it would give us better guarantees that types are correct and expected

Reply:
The specs are built in the loop, I use the type that is returned from GetSchema when looping through instead of the types that are in the arrow array.

Reply:
My bad. Missed that